--- a/Engine/Source/Runtime/ApplicationCore/Private/IOS/IOSAppDelegate.cpp	2025-04-24 15:20:59.251822100 -0500
+++ b/Engine/Source/Runtime/ApplicationCore/Private/IOS/IOSAppDelegate.cpp	2025-05-12 14:28:11.152778700 -0500
@@ -86,6 +86,44 @@
 		TEXT("How long to block main IOS thread to make sure gamethread gets time.\n"),
 		ECVF_Default);
 
+// [CleverTap] Begin
+#if !PLATFORM_TVOS
+
+namespace {
+
+struct FRemoteNotificationMomento
+{
+  UNUserNotificationCenter* Center{ nil };
+  UNNotificationResponse* Response{ nil };
+};
+
+FRemoteNotificationMomento SavedRemoteNotificationMomento;
+
+void HandleSavedRemoteNotificationIfAvailable()
+{
+	if (SavedRemoteNotificationMomento.Response == nil)
+	{
+		return;
+	}
+
+	void (^EmptyHandler)() = ^() {
+		// Intentional no-op
+	};
+
+	id Delegate = [UNUserNotificationCenter currentNotificationCenter].delegate;
+	[Delegate userNotificationCenter:SavedRemoteNotificationMomento.Center
+	  didReceiveNotificationResponse:SavedRemoteNotificationMomento.Response
+	           withCompletionHandler:EmptyHandler];
+
+	[SavedRemoteNotificationMomento.Center release];
+	[SavedRemoteNotificationMomento.Response release];
+	SavedRemoteNotificationMomento = FRemoteNotificationMomento{};
+}
+
+}
+
+#endif
+// [CleverTap] End
 
 static void SignalHandler(int32 Signal, struct __siginfo* Info, void* Context)
 {
@@ -372,6 +410,12 @@
 	});
 #endif
 
+	// [CleverTap] Begin
+#if !PLATFORM_TVOS
+	HandleSavedRemoteNotificationIfAvailable();
+#endif
+	// [CleverTap] End
+
     while( !IsEngineExitRequested() )
 	{
         if (self.bIsSuspended)
@@ -1639,6 +1683,16 @@
 	}
 	
 	HandleReceivedNotification(response.notification);
+
+// [CleverTap] Begin
+#if !PLATFORM_TVOS
+	if (!([IOSAppDelegate GetDelegate].bEngineInit))
+	{
+		SavedRemoteNotificationMomento.Center = [center retain];
+		SavedRemoteNotificationMomento.Response = [response retain];
+	}
+#endif
+// [CleverTap End]
 	
 	completionHandler();
 }
